packages/openactive-integration-tests/test/helpers/flow-stages/fetch-opportunities.js

```diff
@@ -0,0 +1,177 @@
+   * the same request may be made over and over.
+   *
+   * @type {Map<number, Promise<ChakramResponse>>}
+   */
+  const reusableOpportunityResponsePromises = new Map();
+
+  /**
+   * Test interface responses - one for each criteria. Only contains id
+   */
+  const testInterfaceOpportunities = await Promise.all(orderItemCriteriaList.map(async (orderItemCriteriaItem, i) => {
+    // If an opportunity is available for reuse, return it
+    const hasReuseKey = !_.isNil(orderItemCriteriaItem.opportunityReuseKey);
+    if (hasReuseKey && reusableOpportunityResponsePromises.has(orderItemCriteriaItem.opportunityReuseKey)) {
+      return await reusableOpportunityResponsePromises.get(orderItemCriteriaItem.opportunityReuseKey);
+    }
+
+    const sellerKey = orderItemCriteriaItem.seller || 'primary';
+    const seller = SELLER_CONFIG[sellerKey];
+    const opportunityResponsePromise = USE_RANDOM_OPPORTUNITIES
+      ? requestHelper.getRandomOpportunity(
+        orderItemCriteriaItem.opportunityType,
+        orderItemCriteriaItem.opportunityCriteria,
+        i,
+        seller['@id'],
+        seller['@type'],
+      ) : requestHelper.createOpportunity(
+        orderItemCriteriaItem.opportunityType,
+        orderItemCriteriaItem.opportunityCriteria,
+        i,
+        seller['@id'],
+        seller['@type'],
+      );
+
+    // If this opportunity can be reused, store it
+    if (hasReuseKey) {
+      reusableOpportunityResponsePromises.set(orderItemCriteriaItem.opportunityReuseKey, opportunityResponsePromise);
+    }
+
+    return await opportunityResponsePromise;
+  }));
+  return testInterfaceOpportunities;
+}
+
+/**
+ * Generally the first FlowStage which is run in a test.
+ * It gets/creates some opportunities from the BookingSystem's opportunity feed.
+ * These opportunities will then be used by subsequent stages.
+ * It makes use of the OpportunityFeedUpdate FlowStage as both stages fetch opportunities
+ * from the opportunities feed.
+ */
+const FetchOpportunitiesFlowStage = {
+  /**
+   * @param {object} args
+   * @param {OpportunityCriteria[]} args.orderItemCriteriaList
+   * @param {string} [args.uuid] UUID to use for Order. If excluded, this will
+   *   be generated.
+   * @param {BaseLoggerType} args.logger
+   * @param {RequestHelperType} args.requestHelper
+   */
+  create({ orderItemCriteriaList, uuid, requestHelper, logger }) {
+    return new FlowStage({
+      testName: 'Fetch Opportunities',
+      runFn: async () => await FetchOpportunitiesFlowStage.run({ orderItemCriteriaList, requestHelper }),
+      itSuccessChecksFn(flowStage) {
+        OpportunityFeedUpdateFlowStage.itSuccessChecks({
+          orderItemCriteriaList,
+          getterFn: () => flowStage.getResponse(),
+        });
+      },
+      itValidationTestsFn(flowStage) {
+        OpportunityFeedUpdateFlowStage.itValidationTests({
+          logger,
+          orderItemCriteriaList,
+          getterFn: () => flowStage.getResponse(),
+        });
+      },
```

Comment:
Note that this FlowStage doesn't use the same FlowStageUtils helper functions to generate success checks and validation tests.

That's because this FlowStage is a little more complicated - it doesn't just produce a single OpenActive item, but it produces a lot of data. This is why it has bespoke test functions